src/Microsoft.AspNet.Mvc.Testing/ControllerTestHelper.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.FeatureModel;
+using Microsoft.AspNet.Http.Core;
+using Microsoft.AspNet.Http.Interfaces;
+using Microsoft.AspNet.Routing;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.DependencyInjection.Fallback;
+
+namespace Microsoft.AspNet.Mvc.Testing
+{
+	public static class ControllerTestHelper
+	{
+		public static Controller Initialize(Controller controller, ControllerTestHelperContext context)
+		{
+			return Initialize(controller, new ActionDescriptor(), context);
+			// Question: Can we build (fully or partially) ActionDescriptor() from tyep information??
+		}
+
+		private static DefaultHttpContext CreateHttpContext(IServiceProvider serviceProvider)
+		{
+			var featureCollection = new FeatureCollection();
+			var featureContext = new FeatureContext();
+
+			featureCollection.Add(typeof(IHttpRequestLifetimeFeature), new FeatureContext());
+			featureCollection.Add(typeof(IHttpRequestFeature), new FeatureContext());
+
+			var httpContext = new DefaultHttpContext(featureCollection)
+			{
+				RequestServices = serviceProvider
+			};
+
+			return httpContext;
+		}
+
+		private static Controller Initialize(Controller controller, ActionDescriptor actionDescriptor,
+			ControllerTestHelperContext helperContext)
+		{
+			helperContext = helperContext ?? new ControllerTestHelperContext();
+
+			helperContext.RouteData = helperContext.RouteData ?? new RouteData();
+			helperContext.ServiceCollection = helperContext.ServiceCollection ?? new ServiceCollection();
+			helperContext.ServiceProvider = helperContext.ServiceProvider ??
+				helperContext.ServiceCollection.BuildServiceProvider();
+			helperContext.HttpContext = CreateHttpContext(helperContext.ServiceProvider);
+
+			controller.ActionContext =
+				new ActionContext(helperContext.HttpContext, helperContext.RouteData, actionDescriptor);
+			controller.Url = new MockUrlHelper();
+
+			helperContext.HttpContext.Request.Form = helperContext.RequestFormCollection;
+
+			if (helperContext.RequestAborted)
+			{
+				helperContext.HttpContext.Abort();
+			}
+
+			((MockUrlHelper)controller.Url).OnAction
+				= (action, controllerName, values, protocol, host, fragment)
+				=>
+				{
+					helperContext.RequestScheme = protocol;
+					helperContext.RequestHost = host;
+					helperContext.RequestFragment = fragment;
```

Comment:
I updated helperContext here to simplify the OnUrlAction callback. not sure if this a good design or not. Otherwise, the signature of OnUrlAction will be Func<string, string, object, string, string, string, string> OnUrlAction. I think this is too long for users.